modules/claims_api/lib/benefits_documents_service/token.rb

```diff
@@ -0,0 +1,99 @@
+# frozen_string_literal: true
+
+require 'auth_token_service/token_service'
+
+module ClaimsApi
+  module BenefitsDocumentsService
+    class Token
+      def initialize(client)
+        @client = client
+        @token_service = AuthTokenService::TokenService
+        @tolerance = @client.auth_token_tolerance
+      end
+
+      # gets new or retrieves existing
+      def get_token(service_name)
+        if @token_service.access_token_exists?(service_name, @tolerance)
+          token = @token_service.get_access_token(service_name)
+        else
+          token = get_new_token
+
+          @token_service.set_access_token(service_name, token)
+        end
+
+        token
+      end
+
+      def get_new_token
+        body = {
+          grant_type: @client.auth_grant_type,
+          client_assertion_type: @client.auth_assertion_type,
+          client_assertion: assertion,
+          scope: @client.auth_scopes
+        }
+
+        res = auth_client.post(
+          meta[:token_endpoint],
+          URI.encode_www_form(body),
+          {
+            Accept: 'application/json',
+            'Content-Type': 'application/x-www-form-urlencoded'
+          }
+        )
+
+        log_data(res)
+
+        res.body[:access_token]
+      end
+
+      def meta
+        raise StandardError, 'client_base_url missing' if @client.auth_client_base_url.blank?
+
+        auth_client.get(@client.auth_client_base_url).body
+      end
+
+      def assertion
+        raise StandardError, 'client_id missing' if @client.auth_client_id.blank?
+
+        raise StandardError, 'client_secret missing' if @client.auth_private_key_path.blank?
+
+        rsa_instance = build_rsa_instance
+
+        payload = {
+          aud: "#{meta[:issuer]}/v1/token",
+          jti: SecureRandom.uuid,
+          iss: @client.auth_client_id,
+          sub: @client.auth_client_id,
+          iat: Time.now.to_i,
+          exp: 5.minutes.from_now.to_i
+        }
+        JWT.encode(payload, rsa_instance, @client.auth_alg.to_s)
+      end
+
+      private
+
+      def build_rsa_instance
+        private_key_path = @client.auth_private_key_path
+        key = File.exist?(private_key_path) ? File.read(private_key_path) : private_key_path
+
+        OpenSSL::PKey::RSA.new(key)
+      end
+
+      def auth_client
+        Faraday.new do |f|
+          f.request :json
+          f.response :raise_error
+          f.response :json, parser_options: { symbolize_names: true }
+          f.adapter Faraday.default_adapter
+        end
+      end
+
+      def log_data(response)
+        expiration = response.body[:expires_in]
+        time_data = Time.zone.now.strftime('%A %B %m @ %I:%M')
+        msg = "Response from Okta to get new token #{time_data} - TTL: #{expiration}"
```

Comment:
DataDog logs will give us the time so we probably don't _need_ this, but I'm fine leaving it.